ext/pdo/pdo_stmt.c

```diff
@@ -772,90 +745,81 @@ static bool do_fetch(pdo_stmt_t *stmt, zval *return_value, enum pdo_fetch_type h
 			break;
 
 		case PDO_FETCH_OBJ:
-			object_init_ex(return_value, ZEND_STANDARD_CLASS_DEF_PTR);
+			ce = zend_standard_class_def;
+			object_init(return_value);
 			break;
 
 		case PDO_FETCH_CLASS:
+			ce = stmt->fetch.cls.ce;
 			if (flags & PDO_FETCH_CLASSTYPE) {
-				zval val;
-				zend_class_entry *cep;
-
-				old_ce = stmt->fetch.cls.ce;
-				ZVAL_COPY_VALUE(&old_ctor_args, &stmt->fetch.cls.ctor_args);
-				old_arg_count = stmt->fetch.cls.fci.param_count;
-				do_fetch_opt_finish(stmt, 0);
-
-				fetch_value(stmt, &val, column_index_to_fetch++, NULL);
-				if (Z_TYPE(val) != IS_NULL) {
-					if (!try_convert_to_string(&val)) {
-						return 0;
-					}
-					if ((cep = zend_lookup_class(Z_STR(val))) == NULL) {
-						stmt->fetch.cls.ce = ZEND_STANDARD_CLASS_DEF_PTR;
-					} else {
-						stmt->fetch.cls.ce = cep;
-					}
+				zval ce_name_from_column;
+				fetch_value(stmt, &ce_name_from_column, column_index_to_fetch++, NULL);
+				/* This used to use try_convert_to_string() which would silently support integers, floats, null
+				 * even if any such value could not generate a valid class name, as no class was found it would
+				 * then proceed to use stdClass */
+				// TODO Raise PDO implementation error when the column name is not a string
+				if (Z_TYPE(ce_name_from_column) == IS_STRING) {
+					ce = zend_lookup_class(Z_STR(ce_name_from_column));
 				}
-
-				do_fetch_class_prepare(stmt);
-				zval_ptr_dtor_str(&val);
-			}
-			ce = stmt->fetch.cls.ce;
-			/* TODO: Make this an assertion and ensure this is true higher up? */
-			if (!ce) {
-				/* TODO Error? */
-				pdo_raise_impl_error(stmt->dbh, stmt, "HY000", "No fetch class specified");
-				return 0;
+				/* Use default CE if present */
+				if (ce == NULL) {
+					ce = zend_standard_class_def;
+				}
+				zval_ptr_dtor(&ce_name_from_column);
+			} else {
+				/* This can happen if the fetch flags are set via PDO::setAttribute()
+				 * $pdo->setAttribute(PDO::ATTR_DEFAULT_FETCH_MODE, PDO::FETCH_CLASS);
+				 * See ext/pdo/tests/bug_38253.phpt */
+				if (UNEXPECTED(ce == NULL)) {
+					pdo_raise_impl_error(stmt->dbh, stmt, "HY000", "No fetch class specified");
+					goto in_fetch_error;
+				}
+				ctor_arguments = stmt->fetch.cls.ctor_args;
 			}
-			if ((flags & PDO_FETCH_SERIALIZE) == 0) {
-				if (UNEXPECTED(object_init_ex(return_value, ce) != SUCCESS)) {
-					return 0;
+			ZEND_ASSERT(ce != NULL);
+			if (flags & PDO_FETCH_SERIALIZE) {
+				if (!ce->unserialize) {
+					/* As this option is deprecated we do not bother to mention the class name. */
+					pdo_raise_impl_error(stmt->dbh, stmt, "HY000", "cannot unserialize class");
+					goto in_fetch_error;
 				}
-				if (!stmt->fetch.cls.fci.size) {
-					if (!do_fetch_class_prepare(stmt)) {
-						zval_ptr_dtor(return_value);
-						return 0;
-					}
+			} else {
+				if (UNEXPECTED(object_init_ex(return_value, ce) != SUCCESS)) {
+					return false;
```

Comment:
needs to go to in_fetch_error
One could wonder if it's not better to rename this function to do_fetch_real, and have a do_fetch function that calls do_fetch_real but sets the statement in_fetch to true before the call and resets it afterwards, then you can avoid all the goto mess.